### DISCUSSION TITLE
Extending TipTap

### DISCUSSION THREAD

**Discussion by richdynamix** [awcodes/filament-tiptap-editor#100] [Q&A] [answered]
Is there a way to extend the TipTap editor to allow some custom HTML? 

During an import from Bard in Statamic to TipTap in Filament, our `figure` element is just being set as a `img` tag as it seems the figure is not supported in the default setup. This is also noticed when adding an image via the Curator picker, the image is rendered as just a `img` with an alt tag, but doesn't add a figure with caption when the image has it saved correctly. 

I would like to add additional media types such as twitter embedded tweets, currently this is just a blockquote with the accompanying script tag, but as TipTap doesn't recognise the script tag, it is stripped out from the saved HTML.

**Comment by awcodes** [comment, marked as answer]
It's not documented, since it's still a work in progress, but you can make your own extensions. Give it a go [https://gist.github.com/awcodes/fb0425c4a2afd0fc8c3318784d9a53fb](https://gist.github.com/awcodes/fb0425c4a2afd0fc8c3318784d9a53fb)

**Reply by richdynamix** [reply]
Brilliant! Thank you, I will give this a go.

**Reply by christopher87** [reply]
Hello,

I have tried to create my own extension with your example. But I guess it didn't find the JS file, because I have an error that says :

`module.esm.js:421 Alpine Expression Error: editor(...).chain(...).focus(...).setPhrase is not a function`

I have just copy/paste your code, change extension name on each file, and change command name ("setPharse" instead "setHurdle")

Did I miss something ?

I would like to create a custom menu with submenus like Hurdle, on click on submenu it displays a sentence.
I would like to reproduce this thing I have done on TinyMCE :

![image](https://i.ibb.co/0t6kqdv/Screenshot-2023-08-23-at-18-20-16.jpg)

Thanks in advance.

**Reply by awcodes** [reply]
Are you on v2 or v3?

**Reply by christopher87** [reply]
I use v2, I am afraid that v3 (Filament) is not stable enough for production and I don't know if all the plugins I use are compatible with v3.

**Reply by awcodes** [reply]
I wouldn't recommend v3 for production yet, either.

I can't think of anything that would have changed in v2 that would keep it from working. 

Can you share all the code you implemented in a gist?
